lib/puppet/type/keystore.rb

```diff
@@ -0,0 +1,55 @@
+Puppet::Type.newtype(:keystore) do
+  desc 'generates an empty pkcs12 keystore'
+
+  ensurable
+
+  newparam(:keystore, :namevar => true) do
+    desc "Path to the keystore"
+    isrequired
+  end
+
+  newparam(:password_file) do
+    desc "Path to file containing the keystore password"
+    isrequired
+  end
+
+  newparam(:owner, parent: Puppet::Type::File::Owner) do
+    desc "Specifies the owner of the keystore. Valid options: a string containing a username or integer containing a uid."
+  end
+
+  newparam(:group, parent: Puppet::Type::File::Group) do
+    desc "Specifies a permissions group for the keystore. Valid options: a string containing a group name or integer containing a gid."
+  end
+
+  newparam(:mode, parent: Puppet::Type::File::Mode) do
+    desc "Specifies the permissions mode of the keystore. Valid options: a string containing a permission mode value in octal notation."
+  end
+
+  autorequire(:file) do
+    [self[:password_file]]
+  end
```

Comment:
Confirming I understood this correctly:

The Keystore will have a `password_file` (L6-9 above) so that the interface will include:

```puppet
keystore { $some_keystore:
  password_file => $some_password_file_path,
}
```

And this autorequire is silently similar to:

```puppet
keystore { $some_keystore:
  password_file => $some_password_file_path,
  require => File[$some_password_file_path],
}
```

?

Reply:
Correct.

Reply:
It turns out this is extremely similar to the example of autorequires in the type development documentation.

The one difference called out there: "An important difference between automatic and explicit relationships is that automatic relationships do not require the other resources to exist, while explicit relationships do."

Reply:
The pure Puppet equivalent is more like a [collector](https://puppet.com/docs/puppet/7/lang_collectors.html). Untested (so there may be typos) but roughly:
```puppet
keystore { $some_keystore:
  password_file => $some_password_file_path,
}
File <| title == $some_password_file_path |> -> Keystore[$some_keystore]
```